utility/service/service.go

```diff
@@ -70,17 +96,102 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
 		return nil, fmt.Errorf("Error admitting relay: %w", err)
 	}
 
-	// TODO: implement Persist Relay
-	// TODO: implement execution
-	// TODO: implement state maintenance
-	// TODO: validate the response from the node?
-	// TODO: (QUESTION) Should we persist SignedRPC?
-	return nil, nil
+	response, err := s.executeRelay(relay)
+	if err != nil {
+		return nil, fmt.Errorf("Error executing relay: %w", err)
+	}
+
+	// DISCUSS: should we validate the response from the node?
+	relayDigest, shouldStore, err := s.hasCollision(relay, response)
+	if err != nil {
+		return nil, fmt.Errorf("Error calculating relay service digest: %w", err)
+	}
+	if !shouldStore {
+		return response, nil
+	}
+
+	height := s.GetBus().GetConsensusModule().CurrentHeight()
+	writeCtx, err := s.GetBus().GetPersistenceModule().NewRWContext(int64(height))
+	if err != nil {
+		return nil, fmt.Errorf("Error getting a write context to update token usage for application %s: %w", relay.Meta.ApplicationAddress, err)
+	}
+	defer writeCtx.Release()
+
+	// DISCUSS: should we extend/use UnitOfWork for updating/retrieving token usage?
+	if err := writeCtx.RecordRelayService(relay.Meta.ApplicationAddress, relayDigest, relay, response); err != nil {
+		return nil, fmt.Errorf("Error recording service proof for application %s: %w", relay.Meta.ApplicationAddress, err)
+	}
+
+	return response, nil
+}
+
+// hasCollision returns:
+//  1. The signed digest of a relay/response pair,
+//  2. Whether there was a collision for the specific chain (i.e. should the service proof be stored for claiming later)
+func (s *servicer) hasCollision(relay *coreTypes.Relay, response *coreTypes.RelayResponse) (digest []byte, collides bool, err error) {
```

Comment:
s/collides/isReplay

Reply:
Thank you for the review. I updated to `isRelayVolumeApplicable` as a proposed change since it seems to be the intention of the computation here, but please let me know if `isReplay` is a better fit.